src/ArraySet/ArraySet.ts

```diff
@@ -0,0 +1,203 @@
+        pipe(
+            set,
+            Array.contains(elem, equalityComparer),
+            Bool.match({
+                true: set,
+                false: unsafe([elem, ...set]),
+            })
+        )
+
+/**
+ * Remove an element from the `ArraySet`, using the given `EqualityComparer` if provided.
+ * If the element does not exist, will return an `ArraySet` with the same elements.
+ *
+ * @group Transformations
+ *
+ * @example
+ * pipe(
+ *   ArraySet.of(['A', 'B', 'C']),
+ *   ArraySet.remove('B')
+ * ) // => ['A', 'C']
+ */
+export const remove =
+    <T>(
+        elem: NoInfer<T>,
+        { equals }: EqualityComparer<NoInfer<T>> = EqualityComparer.Default
+    ) =>
+    (set: ArraySet<T>): ArraySet<T> =>
+        pipe(
+            set,
+            Array.filter(el => !equals(el, elem)),
+            unsafe
+        )
+
+/**
+ * Convert an existing `Array` into an `ArraySet`, using the given `EqualityComparer`.
+ * Analogous to `new Set([...])`.
+ *
+ * @group Constructors
+ */
+export const ofArray = <T>(
+    arr: readonly T[],
+    equalityComparer?: EqualityComparer<T>
+): ArraySet<T> => pipe(arr, Array.uniq(equalityComparer), unsafe)
+
+/**
+ * Alias for {@link ofArray}
+ *
+ * @group Constructors
+ */
+export const of = ofArray
+
+/**
+ * Create a new `ArraySet` from the result of merging two `ArraySet`s and keeping only
+ * the unique values, according to the given `EqualityComparer` (or default/triple-equals
+ * if not specified).
+ *
+ * @group Set Arithmetic
+ *
+ * @example
+ * pipe(
+ *   ArraySet.of([1, 2, 3, 4]),
+ *   ArraySet.union(ArraySet.of([3, 4, 5]))
+ * ) // => [1, 2, 3, 4, 5]
+ */
+export const union =
+    <T>(set2: ArraySet<T>, equalityComparer?: EqualityComparer<T>) =>
+    (set1: ArraySet<T>): ArraySet<T> =>
+        pipe([...set1, ...set2], Array.uniq(equalityComparer), unsafe)
+
+/**
+ * Create a new `ArraySet` by keeping only elements that are contained in both `ArraySet`s.
+ * Uses the given `EqualityComparer` if specified; defaults to reference/triple-equals.
+ *
+ * @group Set Arithmetic
+ *
+ * @example
+ * pipe(
+ *   ArraySet.of([1, 2, 3, 4]),
+ *   ArraySet.intersect(ArraySet.of([3, 4, 5]))
+ * ) // => [3, 4]
+ */
+export const intersect =
+    <T>(set2: ArraySet<T>, equalityComparer?: EqualityComparer<T>) =>
+    (set1: ArraySet<T>): ArraySet<T> => {
+        const sets = [set1, set2] as const
+
+        const intersection: T[] = []
+
+        const smallerIdx = set1.length > set2.length ? 1 : 0
+        const biggerIdx = smallerIdx === 1 ? 0 : 1
+
+        const smaller = sets[smallerIdx]
+        const bigger = sets[biggerIdx]
+
+        for (const elem of bigger) {
+            pipe(
+                smaller,
+                Array.contains(elem, equalityComparer),
+                Bool.match({
+                    true: () => {
+                        intersection.push(elem)
+                    },
+                    false: undefined,
+                })
+            )
```

Comment:
⛏️ / 🎨 I think an `if` would probably be simpler here. e.g.
```ts
if (pipe(smaller, Array.contains(elem, equalityComparer))) {
    intersection.push(elem)
}
```

I'm also wondering if we could just do a filter instead here 🤔 
```ts
return pipe(
    bigger, 
    Array.filter(elem => 
        pipe(smaller, Array.contains(elem, equalityComparer))
    ),
    unsafe
)
```

Reply:
Hmm.. yeah the reason I went for a `Bool.match` function is because I find it "cleaner" to read when I don't have to mix expressions and statements.. When everything can just go in the `pipe` expression in a series of operations, it feels easy to reason about what's going on.

However, you raise an interesting point about the filter. That actually made me think, I think we can make it more efficient by actually _removing_ the element from the `smaller` array when we find a match, so that subsequent `contains` iterations have one less element to search through in the worst case.

Reply:
Works for me :+1: Thanks!